Allow completion handlers to be registered before binding

austinzheng opened the pull request:
Changes:
- Completion handlers registered before binding are now queued instead of being silently dropped

To do:
- [x] Tests

austinzheng commented:
Getting this (or something with equivalent functionality) in is a prerequisite for the download-a-Realm-asynchronously API we're planning.

austinzheng commented:
On hold until I can pull the test changes out.

bdash commented on src/sync/sync_session.cpp:
```diff
@@ -608,28 +643,35 @@ bool SyncSession::can_wait_for_network_completion() const
     std::unique_lock<std::mutex> lock(m_state_mutex);
-    // FIXME: instead of dropping the callback if we haven't yet `bind()`ed,
-    // save it and register it when the session `bind()`s.
     if (can_wait_for_network_completion()) {
```
Rather than switching on the session state, can we instead just delegate to a function on the state?

bdash commented on src/sync/sync_session.hpp:
```diff
@@ -263,6 +266,17 @@ class SyncSession : public std::enable_shared_from_this<SyncSession> {
+
+    // For storing wait-for-completion requests if the session isn't yet ready to handle them.
+    struct CompletionWaitPackage {
+        CompletionWaitType type;
```
Since the only thing we use this enum for is to determine which function to call, we could instead directly store a member function pointer and avoid having to switch on the enum.

austinzheng commented:
I addressed @bdash's feedback. I want to see if I can get rid of `wait_for_upload_completion_blocking()` completely, since it is only used in one place and by a single test.

bdash commented on src/sync/sync_session.hpp:
```diff
@@ -49,6 +49,7 @@ namespace sync {
 class Session;
 }
 
+using SessionWaiterPointer = void(sync::Session::*)(std::function<void(std::error_code)>);
```
Since this is only used internally it'd be better to put this inside the class, alongside the `private` members.

bdash commented on src/sync/sync_session.hpp:
```diff
@@ -211,6 +214,7 @@ class SyncSession : public std::enable_shared_from_this<SyncSession> {
     // }
 
     bool can_wait_for_network_completion() const;
+    bool can_queue_for_network_completion() const;
```
These are both gone now.

bdash commented on tests/results.cpp:
```diff
@@ -1042,7 +1042,14 @@ TEST_CASE("notifications: sync") {
+            cv.notify_one();
+        });
+        std::unique_lock<std::mutex> lock(wait_mutex);
+        cv.wait(lock);
```
`std::condition_variable::wait` can return spuriously, so there's the potential for this code to continue executing prior to the upload completing. I think you you can use the variant that takes a predicate, combined with the upload completion handler setting a flag, to avoid problems due to spurious unblocking.

austinzheng commented:
I've rebased onto master and updated this PR again, @bdash. There was also an issue with a changeset cooker variable assignment that I had to fix for things to compile.

bdash commented on tests/sync/session/wait_for_completion.cpp:
```diff
@@ -0,0 +1,256 @@
+        REQUIRE(handler_called == false);
+        // Spin the loop many times.
+        std::atomic<int> spin_count(0);
+        EventLoop::main().run_until([&] { spin_count++; return spin_count > 100; });
```
This single section of the test appears to run for over a second, predominantly due to this spinning. This pattern is repeated a few times in these new tests, adding around 5 seconds to the total runtime of all object store tests. That doubles how long it takes the object store tests to run. Is it necessary to spin for so long?
